netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java

```diff
@@ -112,168 +97,252 @@ public static ChannelHandler serverTls(SSLEngine sslEngine) {
+       * because the initializer is not executed in the calling thread but on channelRegistered
+       * in the eventloop thread. So there is a race condition between threads immediately writing
+       * to the channel (from outside the eventloop) and the pipeline being set up i.e. a
+       * write task (writing to the HeadContext) may be created before the pipeline is set up.
+       */
+      if (handlers != null) {
+        ctx.pipeline().addFirst(handlers);
+        handlers = null;
       }
-    };
+      super.channelRegistered(ctx);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+      failBufferedAndClose(ctx);
+      super.channelInactive(ctx);
+    }
 
-    return new Negotiation() {
-      @Override
-      public ChannelInitializer<Channel> initializer() {
-        return initializer;
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+        throws Exception {
+      /**
+       * This check is necessary as there is a race condition between threads writing to the
+       * channel and the handler removing itself from the pipeline (once all buffered messages
+       * are written) i.e. channel.write(msg) may add a write task to the eventloop that invokes
+       * write on the ChannelHandlerContext for this handler, but before the write task is executed
+       * this handler may remove itself from the pipeline. For details have a look at the
+       * implementation of AbstractChannelHandlerContext#write(Object msg, ChannelPromise promise).
+       */
+      if (bufferedWrites == null) {
+        super.write(ctx, msg, promise);
+      } else {
+        bufferedWrites.add(new ChannelWrite(msg, promise));
       }
+    }
 
-      @Override
-      public ListenableFuture<Void> completeFuture() {
-        return completionHandler.getUpgradeFuture();
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+      /**
+       * Don't propagate flush while there are still messages in the buffer.
+       * This is required to make buffering with the SslHandler work, as otherwise
+       * TLS handshakes "randomly" timeout. The SslHandler doesn't like flushes while the
+       * protocol is not negotiated...
+       */
+      if (bufferedWrites == null) {
+        ctx.flush();
       }
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise future) throws Exception {
+      failBufferedAndClose(ctx);
+    }
 
-      @Override
-      public void onConnected(Channel channel) {
-        // Trigger the HTTP/1.1 plaintext upgrade protocol by issuing an HTTP request
-        // which causes the upgrade headers to be added
-        DefaultHttpRequest upgradeTrigger =
-            new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/");
-        channel.writeAndFlush(upgradeTrigger);
+    void failBufferedAndClose(ChannelHandlerContext ctx) {
```

Comment:
Maybe make these `protected` since they are intended to be called by a subclass? I know everything is in the same package so package-private works too, but maybe we should move some/all of these to top-level classes? WDYT?